types/tx/middleware.go

```diff
@@ -0,0 +1,33 @@
+package tx
+
+import (
+	context "context"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// RequestSimulateTx is the request type for the tx.Handler.RequestSimulateTx
+// method.
+type RequestSimulateTx struct {
+	TxBytes []byte
+}
+
+// ResponseSimulateTx is the response type for the tx.Handler.RequestSimulateTx
+// method.
+type ResponseSimulateTx struct {
+	GasInfo sdk.GasInfo
+	Result  *sdk.Result
+}
+
+// TxHandler defines the baseapp's CheckTx, DeliverTx and Simulate respective
+// handlers. It is designed as a middleware stack.
+type Handler interface {
+	CheckTx(ctx context.Context, tx sdk.Tx, req abci.RequestCheckTx) (abci.ResponseCheckTx, error)
+	DeliverTx(ctx context.Context, tx sdk.Tx, req abci.RequestDeliverTx) (abci.ResponseDeliverTx, error)
+	SimulateTx(ctx context.Context, tx sdk.Tx, req RequestSimulateTx) (ResponseSimulateTx, error)
```

Comment:
in most cases these may have a lot of overlap. Have you thought about merging checkTx and SimulateTx to avoid an extra tid bit of boiler plate

Reply:
True, but there are still some differences between the two: SimulateTx sets an InfiniteGasMeter, and doesn't verify sigs etc.

I guess the alternate high-level design is one method `RunTx` instead of the 3, and we pass around a flag `mode = deliver/check/recheck/simulate`. Probably less duplicate code, but more checks like `if checkTx && !simulateTx {do...}`. Happy to hear opinions.

Reply:
I think the motivation with separating them is to push implement to do the correct thing for each case. There can still be significant code reuse under the hood

Reply:
We could consider a common `Response` return type for all three methods - that would simplify code reuse a bit.

`Response` could either be something like `ResponseSimulateTx` as above, or all of the fields flattened into the struct as in `ResponseDeliverTx` and `ResponseCheckTx` which are otherwise identical.

CheckTx will have priority, however, so this is likely premature